crates/sui-rosetta/src/state.rs

```diff
@@ -0,0 +1,370 @@
+                    "Unexpected error, cannot find the latest block.",
+                )
+            })
+            .cloned()
+    }
+
+    fn genesis_block_identifier(&self) -> BlockIdentifier {
+        BlockIdentifier {
+            index: 0,
+            hash: BlockHash([0u8; TRANSACTION_DIGEST_LENGTH]),
+        }
+    }
+
+    async fn oldest_block_identifier(&self) -> Result<BlockIdentifier, Error> {
+        self.blocks
+            .read()
+            .await
+            .first()
+            .map(|b| b.block.block_identifier.clone())
+            .ok_or_else(|| {
+                Error::new_with_msg(
+                    BlockNotFound,
+                    "Unexpected error, cannot find the oldest block.",
+                )
+            })
+    }
+
+    async fn current_block_identifier(&self) -> Result<BlockIdentifier, Error> {
+        self.current_block().await.map(|b| b.block.block_identifier)
+    }
+
+    async fn get_balance_at_block(
+        &self,
+        addr: SuiAddress,
+        block_height: u64,
+    ) -> Result<u128, Error> {
+        if let Some(balances) = self.balance.read().await.get(&addr) {
+            for (blk_idx, balance) in balances.iter().rev() {
+                if blk_idx <= &block_height {
+                    return Ok(*balance);
+                }
+            }
+        };
+        Ok(0)
+    }
+}
+
+impl PseudoBlockProvider {
+    pub fn spawn(state: Arc<AuthorityState>, genesis: &Genesis) -> Self {
+        let genesis = genesis_block(genesis);
+        let genesis_txs = genesis
+            .block
+            .transactions
+            .iter()
+            .flat_map(|tx| tx.operations.clone())
+            .collect();
+        let blocks = Self {
+            blocks: Arc::new(RwLock::new(vec![genesis])),
+            balance: Arc::new(Default::default()),
+        };
+
+        let block_interval = option_env!("SUI_BLOCK_INTERVAL")
+            .map(|i| u64::from_str(i).ok())
+            .flatten()
+            .unwrap_or(2000);
+        let block_interval = Duration::from_millis(block_interval);
+
+        let f = blocks.clone();
+        tokio::spawn(async move {
+            f.update_balance(0, genesis_txs).await.unwrap();
+            loop {
+                if let Err(e) = f.create_next_block(state.clone()).await {
+                    error!("Error creating block, cause: {e:?}")
+                }
+                tokio::time::sleep(block_interval).await;
+            }
+        });
+
+        blocks
+    }
+
+    async fn create_next_block(&self, state: Arc<AuthorityState>) -> Result<(), Error> {
+        let current_block = self.current_block_identifier().await?;
+        let total_tx = state.get_total_transaction_number()?;
+        if total_tx == 0 {
+            return Ok(());
+        }
+        if current_block.index < total_tx {
+            // Make sure we don't hit the query limit.
+            let end = min(4000 + current_block.index, total_tx);
```

Comment:
Why 4000?  Could this be configurable?

Reply:
that is a hardcoded value in [query helpers](https://github.com/MystenLabs/sui/blob/30d2fba3a0471a6addb50e50630978f963d34f05/crates/sui-core/src/query_helpers.rs#L12) 

I don't want to depend on core just for this const, also we will have pagination soon, will replace this with pagination when that's available. Added a TODO.